### PR TITLE
pilot-core: imu temperature and gyrorth calibration + gps double usage

### PR DESCRIPTION
**Pull request by niewim19**
## Description

 - adding imu temperature calibration `tempiumu`
 - correction module more modular
 - correction module initialization flags

## Motivation and Context

Adding `tempimu` calibration type forced a change in `sensc` and `corr` so that it is possible to run only selected calibrations and corrections. It is because `tempimu` calibration must be applied on non-calirbated data and current `accorth` calibration must be done on `tempimu` calibrated data, yet it is no possible to do these calibrations together. 

This problem occurs also with `magiron`/`magmot` calibrations and `accorth`/`accrot` but there it was mitigated by joining calibration procedures. However to tidy it up these calibrations will need splitting and the correct order of calibration enforced in `calibtool` with `corrInitFlags`.

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to change)

## How Has This Been Tested?

- [ ] Already covered by automatic testing.
- [ ] New test added: (add PR link here).
- [x] Tested by hand on: zturn drone

## Checklist:

- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing linter checks and tests passed.
- [x] My changes generate no new compilation warnings for any of the targets.

## Special treatment

- [ ] This PR needs additional PRs to work (list the PRs, preferably in merge-order).
- [ ] I will merge this PR by myself when appropriate.

### CODE REVIEW

File: calib/tempimu.c [hunk 1]
```diff
@@ -58,21 +118,86 @@ static int tempimu_write(FILE *file)
...
+	float refTempStart, refTempEnd;
+	int i;
+
+	for (i = 0; i < 3; i++) {
+		linlsm_init(&lsmAcc[i]);
+		linlsm_init(&lsmGyr[i]);
+	}
+
+	/* Averaging start/end temperature 2x more than usual samples */
+	if (tempimu_imuAvg(DATA_POINT_SAMPLES_PERIOD_US * 2, DATA_POINT_SAMPLES_PERIOD_US, &accelEvt, &gyrEvt) != 0) {
```
Comment (PNieck) [inline calib/tempimu.c:142]: Why are you using samples period, where samples count should be passed it's missleading

File: calib/tempimu.c [hunk 2]
```diff
@@ -14,14 +13,75 @@
...
+
+#define MIN_TEMP_DIFF 3
 
 struct {
 	calib_data_t data;
 } tempimu_common;
 
 
+/* Averages accel temperature over `samples` samples with `usDelay` period. Outputs to `out` in kelvins. Returns 0 on success */
+static int tempimu_imuAvg(unsigned int samples, time_t usDelay, sensor_event_t *accOut, sensor_event_t *gyrOut)
```
Comment (PNieck) [inline calib/tempimu.c:41]: You are calculating average value of all accelerometer and gyroscope data, but only temperature is used. Maybe rename this function to `tempimu_imuAvgTemp` and return only temperature.

**Review by PNieck — CHANGES_REQUESTED** [review]

File: calib/gyrorth.c [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+static int gyrorth_done(void)
+{
+	return EOK;
+}
+
+
+int gyrorth_run(void)
+{
+	printf("This calibration returns default parameters if there are none in calibration file\n");
+	printf("Calibration done!");
```
Comment (PNieck) [inline calib/gyrorth.c:68]: Add message, that this form of calibration is not working for now

**Review by PNieck — CHANGES_REQUESTED** [review]

File: libs/calib/calib.c
[the remark this answers is not in the document]
Comment (niewim19) [reply libs/calib/calib.c:1]: Added type setup in `calib_readFile()`

File: libs/calib/calib.c [hunk 4]
```diff
@@ -442,6 +442,121 @@ static int calib_accorthRead(FILE *file, calib_data_t *cal)
...
+{
+	cal->params.tempimu.refTemp = 300;
+
+	cal->params.tempimu.alfaAcc[0] = 0;
+	cal->params.tempimu.alfaAcc[1] = 0;
+	cal->params.tempimu.alfaAcc[2] = 0;
+
+	cal->params.tempimu.alfaGyr[0] = 0;
+	cal->params.tempimu.alfaGyr[1] = 0;
+	cal->params.tempimu.alfaGyr[2] = 0;
```
[the remark this answers is not in the document]
Comment (niewim19) [reply libs/calib/calib.c:495]: Added comment about calibration process and default values

File: libs/calib/calib.c [hunk 5]
```diff
@@ -532,6 +647,7 @@ void calib_free(calib_data_t *cal)
 
 		case typeMagmot:
 		case typeMotlin:
+		case typeTempimu:
 		default:
 			return;
```
[the remark this answers is not in the document]
Comment (niewim19) [reply libs/calib/calib.c:782]: Split into all non-allocating types, and `default` with error message.

File: libs/sensc/sensc.c [hunk 6]
```diff
@@ -203,7 +203,8 @@ int sensc_imuGet(sensor_event_t *accelEvt, sensor_event_t *gyroEvt, sensor_event
 
 	if (sensc_common.corrEnable == true) {
 
-		corr_mag(magEvt);
+		corr_magiron(magEvt);
+		corr_magmot(magEvt);
```
[the remark this answers is not in the document]
Comment (niewim19) [reply libs/sensc/sensc.c]: This is heavily amended in future commits.

File: libs/sensc/sensc.c [hunk 7]
```diff
@@ -121,30 +121,30 @@ static int sensc_openDescr(const char *path, int typeFlag, int initFlags, int *s
...
-int sensc_init(const char *path, bool corrEnable, int initFlags)
+int sensc_init(const char *path, int corrInitFlags, int sensInitFlags)
 {
 	int err;
 
-	sensc_common.corrEnable = corrEnable;
-	if (sensc_common.corrEnable == true) {
+	sensc_common.corrInitFlags = corrInitFlags;
+	if (sensc_common.corrInitFlags != CORR_ENBL_NONE) {
 		if (corr_init(CORR_ENBL_ALL) != 0) {
```
[the remark this answers is not in the document]
Comment (niewim19) [reply libs/sensc/sensc.c]: This is added later in the commits. At this point the goal was to use any flag as it was not fully implemented yet.

File: calib/tempimu.c (hunk 2 above)
Comment (niewim19) [reply calib/tempimu.c:41]: You pointed correctly - there is discrepancy between the name of function, its description and its usage. 

However this function is used both for temperature and imu readings. So it does average over imu measurements, together with temperature. I would stay with `tempimu_imuAvg`.

**Pull request merged**